### PR TITLE
Report v2: Basic table formatter

### PR DESCRIPTION
**Pull request by paulirish**
It's not very pretty but the basics are there.

* adds thumbnail renderer
* adds URL renderer (just displays as text for now)
* adds table renderer (obviously)
* introduces `AuditResult.details.itemHeaders` for the `thead` cells

screenshot, lol:
image

### CODE REVIEW

File: lighthouse-core/audits/byte-efficiency/unused-css-rules.js [hunk 1]
```diff
@@ -179,14 +179,17 @@ class UnusedCSSRules extends Audit {
       return UnusedCSSRules.mapSheetToResult(indexedSheets[sheetId], pageUrl);
     }).filter(sheet => sheet && sheet.wastedBytes > 1024);
 
+
+    const headings = [
+      {key: 'url', itemType: 'url', text: 'URL'},
```
Comment (patrickhulce) [inline lighthouse-core/audits/byte-efficiency/unused-css-rules.js]: we can override this per row correct? it'd be nice to finally have a nice solution for displaying the inline code preview in the url column

File: lighthouse-core/audits/audit.js [hunk 2]
```diff
@@ -57,6 +57,60 @@ class Audit {
...
+    return headings.map(heading => ({
+      type: 'text',
+      text: heading.text
+    }));
+  }
+
+  /**
+   * @param  {!Audit.Headings} headings
+   * @param  {!Object} results
+   * @return {!DetailsRenderer.DetailsJSON}
```
Comment (ebidel) [inline lighthouse-core/audits/audit.js]: the return obj is slightly different than `DetailsJSON`. Should it be a new type? `TableDetailsJSON` or whatever.

File: lighthouse-core/audits/byte-efficiency/offscreen-images.js [hunk 3]
```diff
@@ -132,15 +132,18 @@ class OffscreenImages extends Audit {
...
       return {
         debugString,
         results,
-        tableHeadings: {
-          preview: '',
-          url: 'URL',
-          totalKb: 'Original',
-          potentialSavings: 'Potential Savings',
-        }
+        headings,
```
Comment (ebidel) [inline lighthouse-core/audits/byte-efficiency/offscreen-images.js]: The `@return` docstring of this method needs to be updated to match.

File: lighthouse-core/audits/audit.js (hunk 2 above)
Comment (ebidel) [inline lighthouse-core/audits/audit.js]: Oh, you have it below :) `DetailsRenderer.TableDetailsJSON`

File: lighthouse-core/report/v2/renderer/details-renderer.js [hunk 4]
```diff
@@ -88,6 +116,33 @@ class DetailsRenderer {
     return element;
   }
 
+
+  /**
+   * @param {!DetailsRenderer.TableDetailsJSON} details
```
Comment (ebidel) [inline lighthouse-core/report/v2/renderer/details-renderer.js]: is `TableDetailsJSON` typedef'd anywhere? not seeing it.

File: lighthouse-core/report/v2/renderer/details-renderer.js [hunk 5]
```diff
@@ -55,6 +61,28 @@ class DetailsRenderer {
...
+
+  /**
+   * @param {!DetailsRenderer.DetailsJSON} obj
+   * @return {!Element}
+   */
+  _renderThumbnail(obj) {
+    const element = this._dom.createElement('img', 'lh-thumbnail');
+    element.src = obj.text.url;
+    // ignore obj.text.mimeType
+    element.alt = 'Image preview';
```
Comment (ebidel) [inline lighthouse-core/report/v2/renderer/details-renderer.js]: Should we also add `element.title = obj.text.url;` so there's a full url when users hover over the image?

File: lighthouse-core/report/v2/renderer/details-renderer.js [hunk 6]
```diff
@@ -88,6 +116,33 @@ class DetailsRenderer {
     return element;
   }
 
+
+  /**
+   * @param {!DetailsRenderer.TableDetailsJSON} details
+   * @return {!Element}
+   */
+  _renderTable(details) {
+    const element = this._dom.createElement('details', 'lh-details', {open: true});
```
Comment (ebidel) [inline lighthouse-core/report/v2/renderer/details-renderer.js]: should we open by default? There could be a  lot of stuff in the table.

File: lighthouse-core/report/v2/renderer/details-renderer.js [hunk 7]
```diff
@@ -88,6 +116,33 @@ class DetailsRenderer {
...
+    if (details.header) {
+      element.appendChild(this._dom.createElement('summary')).textContent = details.header;
+    }
+
+    const tableElem = element.createChild('table', 'lh-table lh-table__multicolumn');
```
Comment (ebidel) [inline lighthouse-core/report/v2/renderer/details-renderer.js]: Do you need `lh-table__multicolumn`? That was an impl detail of v1's handlebars table.

File: lighthouse-core/report/v2/renderer/dom.js [hunk 8]
```diff
@@ -23,6 +23,30 @@ class DOM {
...
+    const instance = this;
+
+    /**
+     * From devtools' DOMExtension.js (minus the 3rd customElementType param)
+     * @param {string} elementName
+     * @param {string=} className
+     * @return {!Element}
+     */
+    self.Element.prototype.createChild = function(elementName, className) {
+      const element = instance.createElement(elementName, className);
```
Comment (ebidel) [inline lighthouse-core/report/v2/renderer/dom.js]: pipe through the `attrs` param too?

File: lighthouse-core/report/v2/renderer/dom.js [hunk 9]
```diff
@@ -23,6 +23,30 @@ class DOM {
...
+      this.appendChild(element);
+      return element;
```
Comment (ebidel) [inline lighthouse-core/report/v2/renderer/dom.js]: nit: `return this.appendChild(element);`

File: lighthouse-core/report/v2/renderer/dom.js [hunk 10]
```diff
@@ -23,6 +23,30 @@ class DOM {
...
+    if (self.Element.prototype.createChild) return;
+    const instance = this;
+
+    /**
+     * From devtools' DOMExtension.js (minus the 3rd customElementType param)
+     * @param {string} elementName
+     * @param {string=} className
+     * @return {!Element}
+     */
+    self.Element.prototype.createChild = function(elementName, className) {
```
Comment (ebidel) [inline lighthouse-core/report/v2/renderer/dom.js]: If we're comfortable with modifying built in prototypes, then I have a bunch of other helpers to add :)

File: lighthouse-core/report/v2/report-styles.css [hunk 11]
```diff
@@ -301,4 +301,15 @@ summary.lh-passed-audits-summary::-webkit-details-marker {
   padding: 3px;
 }
 
+
+.lh-table {
+  --image-preview: 24px;
+}
+
+img.lh-thumbnail {
```
Comment (ebidel) [inline lighthouse-core/report/v2/report-styles.css]: just `.lh-thumbnail`? Seems ok to be tag-agnostic.

File: lighthouse-core/audits/byte-efficiency/unused-css-rules.js (hunk 1 above)
Comment (paulirish) [inline lighthouse-core/audits/byte-efficiency/unused-css-rules.js]: to have an individual row change the itemType? that seems a little weird. whyd we want diff itemtypes in the column?

File: lighthouse-core/report/v2/renderer/details-renderer.js (hunk 5 above)
Comment (paulirish) [inline lighthouse-core/report/v2/renderer/details-renderer.js]: sg

File: lighthouse-core/report/v2/renderer/details-renderer.js (hunk 6 above)
Comment (paulirish) [inline lighthouse-core/report/v2/renderer/details-renderer.js]: nah. was only for development. will remove. :)

File: lighthouse-core/report/v2/renderer/details-renderer.js (hunk 7 above)
Comment (paulirish) [inline lighthouse-core/report/v2/renderer/details-renderer.js]: nope. will nuke.

File: lighthouse-core/report/v2/renderer/dom.js (hunk 10 above)
Comment (paulirish) [inline lighthouse-core/report/v2/renderer/dom.js]: check out https://github.com/ChromeDevTools/devtools-frontend/blob/master/front_end/dom_extension/DOMExtension.js .. we can polyfill whatever from there we want. (or easier: just load that file in the report)

File: lighthouse-core/report/v2/renderer/dom.js (hunk 8 above)
Comment (paulirish) [inline lighthouse-core/report/v2/renderer/dom.js]: i dont think we can because in devtools this method will already exist and it doesnt do attrs

File: lighthouse-core/audits/audit.js [hunk 12]
```diff
@@ -57,6 +57,60 @@ class Audit {
   }
 
   /**
+   * @param  {!Audit.Headings} headings
+   * @return {*}
```
Comment (brendankenny) [inline lighthouse-core/audits/audit.js]: `!Array<string>`?

File: lighthouse-core/audits/audit.js [hunk 13]
```diff
@@ -57,6 +57,60 @@ class Audit {
...
+   */
+  static makeV1TableHeadings(headings) {
+    const tableHeadings = {};
+    headings.forEach(heading => tableHeadings[heading.key] = heading.text);
+    return tableHeadings;
+  }
+
+  /**
+   * @param  {!Audit.Headings} headings
+   * @param  {!Object} results
```
Comment (brendankenny) [inline lighthouse-core/audits/audit.js]: not an object. `!Array<!Object<string, string>>`? It's hard to follow the old `createTable` helper

File: lighthouse-core/audits/audit.js [hunk 14]
```diff
@@ -57,6 +57,60 @@ class Audit {
   }
 
   /**
+   * @param  {!Audit.Headings} headings
```
Comment (brendankenny) [inline lighthouse-core/audits/audit.js]: just one space (here and elsewhere). Is this a vscode setting that needs to be changed?

File: lighthouse-core/audits/audit.js [hunk 15]
```diff
@@ -57,6 +57,60 @@ class Audit {
...
+          text: item[heading.key]
+        };
+      });
+    });
+    return tableRows;
+  }
+
+  /**
+   * @param  {!Audit.Headings} headings
+   * @return {!Array<!DetailsRenderer.DetailsJSON>}
```
Comment (brendankenny) [inline lighthouse-core/audits/audit.js]: this is fine for now, but it won't actually work if we were to turn on type checking on this file. Closure won't be able to see the typedef since it's in another module. We'll have to think of how we want to do that (move report typedefs to their own file?)

File: lighthouse-core/audits/audit.js [hunk 16]
```diff
@@ -57,6 +57,60 @@ class Audit {
...
+  static makeV2TableHeaders(headings) {
+    return headings.map(heading => ({
+      type: 'text',
+      text: heading.text
+    }));
+  }
+
+  /**
+   * @param  {!Audit.Headings} headings
+   * @param  {!Object} results
```
Comment (brendankenny) [inline lighthouse-core/audits/audit.js]: same type as above `results`

File: lighthouse-core/audits/audit.js [hunk 17]
```diff
@@ -97,3 +151,14 @@ class Audit {
...
+
+/** @typedef {{
+ *     key: string,
+ *     itemType: string,
+ *     text: string,
+ * }}
+ */
+Audit.Heading; // eslint-disable-line no-unused-expressions
+
+/** @typedef {!Array<!Audit.Heading>} */
```
Comment (brendankenny) [inline lighthouse-core/audits/audit.js]: is it worth making a typedef for this?

File: lighthouse-core/report/v2/renderer/details-renderer.js [hunk 18]
```diff
@@ -55,6 +61,28 @@ class DetailsRenderer {
...
+    return element;
+  }
+
+  /**
+   * @param {!DetailsRenderer.DetailsJSON} obj
+   * @return {!Element}
+   */
+  _renderThumbnail(obj) {
+    const element = this._dom.createElement('img', 'lh-thumbnail');
+    element.src = obj.text.url;
```
Comment (brendankenny) [inline lighthouse-core/report/v2/renderer/details-renderer.js]: this isn't a thing on `DetailsJSON`

File: lighthouse-core/audits/audit.js [hunk 19]
```diff
@@ -57,6 +57,60 @@ class Audit {
...
+  /**
+   * @param  {!Audit.Headings} headings
+   * @param  {!Object} results
+   * @return {!Array<!DetailsRenderer.DetailsJSON>}
+   */
+  static makeV2TableRows(headings, results) {
+    const tableRows = results.map(item => {
+      return headings.map(heading => {
+        return {
+          type: heading.itemType,
```
Comment (brendankenny) [inline lighthouse-core/audits/audit.js]: why not do this in `_renderTable` rather than in the JSON report?

File: lighthouse-core/audits/byte-efficiency/unused-css-rules.js (hunk 1 above)
Comment (patrickhulce) [inline lighthouse-core/audits/byte-efficiency/unused-css-rules.js]: in all the css-based ones the "url column" doesn't always have a url and it's a code snippet preview instead since it's an inline style. it'd be nice to have the details renderer support that too so the code snippets can say `{type: "code", content: ".inline-style {...}"}` or something

**Comment by ebidel** [comment]
@paulirish how's this coming? We've got a lot of audits that need this guy.

**Comment by paulirish** [comment]
> @paulirish how's this coming? We've got a lot of audits that need this guy.

roger that. on it.

File: lighthouse-core/audits/audit.js (hunk 19 above)
Comment (paulirish) [inline lighthouse-core/audits/audit.js]: shrugz

I'm okay with either.
patrick is fine with either but thinks its slightly better to have all things coming into renderer be our details objects
dgozman also feels the same way, and thinks that if we wanted to conserve the size of the LH object we should shorten `type` to `t` etc. ;) ;) 

so I'm comfortable expanding the objects over here before we deliver the LHR.

Comment (brendankenny) [inline lighthouse-core/audits/audit.js]: > so I'm comfortable expanding the objects over here before we deliver the LHR.

I think with our less rigid details taxonomy now I'm actually more OK with this extra verboseness :)

We can keep an eye on report file size and how much is taken up by `"type": "text"` and figure out some defaults (default to text or default to header itemType) if/when needed

File: lighthouse-core/audits/audit.js (hunk 17 above)
Comment (paulirish) [inline lighthouse-core/audits/audit.js]: Actually yeah. Basically we pass around this typedef and never use the Audit.Heading one directly. Not entirely worth it to nuke this guy, i tried.

**Comment by paulirish** [comment]
PTAL. updated.

File: lighthouse-core/audits/audit.js [hunk 20]
```diff
@@ -57,6 +57,65 @@ class Audit {
...
+   * can be provided: results[x].propName = {type: 'code', text: '...'}
+   * @param {!Audit.Headings} headings
+   * @param {!Array<!Object<string, string>>} results
+   * @return {!Array<!DetailsRenderer.DetailsJSON>}
+   */
+  static makeV2TableRows(headings, results) {
+    const tableRows = results.map(item => {
+      return headings.map(heading => {
+        const value = item[heading.key];
+        if (typeof value === 'object' && value.type) return value;
```
Comment (patrickhulce) [inline lighthouse-core/audits/audit.js]: 🎉

File: lighthouse-core/report/v2/report-styles.css [hunk 21]
```diff
@@ -424,4 +424,14 @@ summary.lh-passed-audits-summary::-webkit-details-marker {
   }
 }
 
+.lh-table {
+  --image-preview: 24px;
```
Comment (patrickhulce) [inline lighthouse-core/report/v2/report-styles.css]: nit: `--image-preview-size` or something?

**Comment by ebidel** [comment]
@brendankenny to you

File: lighthouse-core/report/v2/renderer/details-renderer.js [hunk 22]
```diff
@@ -134,3 +197,21 @@ DetailsRenderer.ListDetailsJSON; // eslint-disable-line no-unused-expressions
  * }}
  */
 DetailsRenderer.CardsDetailsJSON; // eslint-disable-line no-unused-expressions
+
+/** @typedef {{
+ *     type: string,
+ *     header: ({text: string}|undefined),
+ *     items: !Array<!Array<!DetailsRenderer.DetailsJSON>>,
+ *     itemHeaders: !Array<!DetailsRenderer.DetailsJSON>
```
Comment (brendankenny) [inline lighthouse-core/report/v2/renderer/details-renderer.js]: by letting these be `DetailsJSON` we're basically descending into recursive madness again since in theory when you call `this.render` on them they could embed a whole other table (but of course, our css would quickly fall over itself if that sort of thing occurred). This leaves these basically untyped (except our assurance to the compiler that we're correct when we cast in `this.render`)

I'm not sure of the right solution...maybe we could separate top level rendering (anything currently in `this.render`) from more primitive-level rendering (just text, url, and thumbnail as of this PR) in a future PR or something, but we would need to make whatever we did non-annoying.

File: lighthouse-core/report/v2/renderer/details-renderer.js [hunk 23]
```diff
@@ -54,6 +60,32 @@ class DetailsRenderer {
...
+    element.classList.add('lh-text__url');
+    return element;
+  }
+
+  /**
+   * @param {!DetailsRenderer.ThumbnailDetails} value
+   * @return {!Element}
+   */
+  _renderThumbnail(value) {
+    if (/^image/.test(value.mimeType) === false) {
```
Comment (brendankenny) [inline lighthouse-core/report/v2/renderer/details-renderer.js]: make a note of this in a jsdoc description?

File: lighthouse-core/report/v2/renderer/details-renderer.js [hunk 24]
```diff
@@ -54,6 +60,32 @@ class DetailsRenderer {
...
+      return this._dom.createElement('span');
+    }
+
+    const element = this._dom.createElement('img', 'lh-thumbnail');
+    element.src = value.url;
+    element.alt = 'Image preview';
```
Comment (brendankenny) [inline lighthouse-core/report/v2/renderer/details-renderer.js]: best we can do?

File: lighthouse-core/audits/byte-efficiency/uses-responsive-images.js [hunk 25]
```diff
@@ -80,8 +81,7 @@ class UsesResponsiveImages extends ByteEfficiencyAudit {
 
   /**
    * @param {!Artifacts} artifacts
-   * @return {{results: !Array<Object>, tableHeadings: Object,
-   *     passes: boolean=, debugString: string=}}
+   * @return {{results: !Array<Object>, headings: !Audit.Headings, passes: boolean=, debugString: string=}}
```
Comment (brendankenny) [inline lighthouse-core/audits/byte-efficiency/uses-responsive-images.js]: `boolean=` would be `(boolean|undefined)`, but doesn't look like it can be undefined?
`string=` would be `(string|undefined)`

(`=` only works for `@params`...technically it means optional param, which just so happens to manifest as being `undefined`)

File: lighthouse-core/audits/byte-efficiency/uses-request-compression.js [hunk 26]
```diff
@@ -46,7 +46,7 @@ class ResponsesAreCompressed extends ByteEfficiencyAudit {
   /**
    * @param {!Artifacts} artifacts
    * @param {number} networkThroughput
-   * @return {!AuditResult}
+   * @return {{results: !Array<Object>, passes: boolean=, headings: !Audit.Headings, debugString: string=}}
```
Comment (brendankenny) [inline lighthouse-core/audits/byte-efficiency/uses-request-compression.js]: see below

File: lighthouse-core/audits/byte-efficiency/uses-optimized-images.js [hunk 27]
```diff
@@ -63,8 +63,7 @@ class UsesOptimizedImages extends ByteEfficiencyAudit {
 
   /**
    * @param {!Artifacts} artifacts
-   * @return {{results: !Array<Object>, tableHeadings: Object,
-   *     passes: boolean=, debugString: string=}}
+   * @return {{results: !Array<Object>, headings: !Audit.Headings, passes: boolean=, debugString: string=}}
```
Comment (brendankenny) [inline lighthouse-core/audits/byte-efficiency/uses-optimized-images.js]: samsies

File: lighthouse-core/audits/byte-efficiency/offscreen-images.js [hunk 28]
```diff
@@ -95,8 +96,7 @@ class OffscreenImages extends ByteEfficiencyAudit {
 
   /**
    * @param {!Artifacts} artifacts
-   * @return {{results: !Array<Object>, tableHeadings: Object,
-   *     passes: boolean=, debugString: string=}}
+   * @return {{results: !Array<Object>, headings: !Audit.Headings, debugString: string=}}
```
Comment (brendankenny) [inline lighthouse-core/audits/byte-efficiency/offscreen-images.js]: see notes below

File: lighthouse-core/audits/audit.js (hunk 17 above)
Comment (brendankenny) [inline lighthouse-core/audits/audit.js]: if that's the case, what about 
```js
/** @typedef {
 * !Array<{
 *   key: string,
 *   itemType: string,
 *   text: string,
 * }>}
 */
Audit.Headings;
```
and skip `Audit.Heading`?

File: lighthouse-core/audits/audit.js (hunk 20 above)
Comment (brendankenny) [inline lighthouse-core/audits/audit.js]: results `value`s might not just be a `string`, then?

File: lighthouse-core/report/v2/renderer/details-renderer.js (hunk 23 above)
Comment (paulirish) [inline lighthouse-core/report/v2/renderer/details-renderer.js]: done

File: lighthouse-core/report/v2/renderer/details-renderer.js (hunk 24 above)
Comment (paulirish) [inline lighthouse-core/report/v2/renderer/details-renderer.js]: was matching our v1. :)  i changed it to empty string, which seems better, actually.

File: lighthouse-core/audits/audit.js (hunk 20 above)
Comment (paulirish) [inline lighthouse-core/audits/audit.js]: fixed

File: lighthouse-core/audits/byte-efficiency/uses-responsive-images.js (hunk 25 above)
Comment (paulirish) [inline lighthouse-core/audits/byte-efficiency/uses-responsive-images.js]: thanks. made a new typdef for this guy since he was everywhere.

File: lighthouse-core/report/v2/renderer/details-renderer.js (hunk 22 above)
Comment (paulirish) [inline lighthouse-core/report/v2/renderer/details-renderer.js]: > we're basically descending into recursive madness again

are we though?


`detailsjson` only has type and text.. it can't contain other types.
really it and `thumbnail` should include the word "item" as they are ALWAYS leaf nodes.

and the `render()` method should accept all of these leaf & group types. the real bug seems that render's param is typed to a leaf node when we know its not.

Comment (brendankenny) [inline lighthouse-core/report/v2/renderer/details-renderer.js]: `detailsjson` is the base (structural) type that is the superset of all of them. It only has fields all of them have: `type` and optional `text` (really text is there from when they all had `text` and should probably be removed). `render()` takes that and casts according to its `type` to the appropriate type, and we'll take responsibility to make sure that if `type` is set to x, then that object really is of `type` x. It's a hack, but simple (we only have to have humans in the loop checking the top level `details` object), and can easily be reworked once we know the full outlines of the types we need.

Allowing arbitrary nesting means that it's much less simple since we have to guarantee that all the nested items and itemHeaders are really whatever their `type` says. It also means we're implicitly breaking the no-typedef-cycles, just getting around it by casting, which is what I meant by recursive madness.

If you're saying we should have a different type explicitly for leaf nodes to make this better, I think that's great

Comment (paulirish) [inline lighthouse-core/report/v2/renderer/details-renderer.js]: ya makes sense. let's sort this out in voice as i have a few ideas.
